[PyROOT] Migrate RDataFrame, RVec and RTensor pythonizations to Python

guitargeek opened the pull request:
Like this, we are using fewer CPyCppyy implementation details in the pythonizations.

guitargeek commented:
Ah! My bad, I didn't notice that the code was still used!

My `git grepping` didn't give me a git because the names of the C++ function and the name of the function on the Python side is different :slightly_smiling_face: 
https://github.com/root-project/root/blob/master/bindings/pyroot/pythonizations/src/PyROOTModule.cxx#L73
```c++
   {(char *)"MakeNumpyDataFrame", (PyCFunction)PyROOT::MakeNumpyDataFrameImpl, METH_O,
    (char *)"Make RDataFrame from dictionary of numpy arrays"},
```

Anyway, I'll keep this PR open and maybe see later how this function can be re-implemented in pure Python.

vepadulano commented on bindings/pyroot/pythonizations/python/ROOT/_facade.py:
```diff
@@ -9,13 +9,186 @@
+    global byteswap
+
+    if len(byteswap) == 0:
+        ROOT.gInterpreter.Declare(
+            """
+std::string PyROOT_getByteswap()
+{
+#ifdef R__BYTESWAP
+    return "<";
+#else
+    return ">";
+#endif
+}
+"""
```
This is not nice to me. I would much rather we had a function somewhere in ROOT (e.g. core/foundation) that we can call to get this information without asking the interpreter

guitargeek replied:
You are right, it's not optimal. However, I would like to keep this PR restricted to the PyROOT code area.

Therefore, I have put back the `GetEndianess()` function in the C++ `libROOTPythonizations` for now, and then we can think later how to "upstream" this to ROOT.

vepadulano commented on bindings/pyroot/pythonizations/python/ROOT/_facade.py:
```diff
@@ -9,13 +9,186 @@
+    # Add PyObject (dictionary) holding RVecs to data source
+    code = "ROOT::Internal::RDF::MakeNumpyDataFrame("
+    code += f"reinterpret_cast<PyObject*>({address_prefix}{id(pyvecs)}), "
+
+    def write_vec_code(key, value):
+        # Get name of key
+        if not isinstance(key, str):
+            raise RuntimeError("Object not convertible: Dictionary key is not convertible to a string.")
+
+        # Convert value to RVec and attach to dictionary
+        pyvec = ROOT.VecOps.AsRVec(value)
+        if not pyvec:
+            raise RuntimeError("Object not convertible: Dictionary entry " + key + " is not convertible with AsRVec.")
+
+        pyvecs[key] = pyvec
+
+        # Add pairs of column name and associated RVec to signature
+        vectype = type(pyvec).__cpp_name__
+        vecaddress = cppyy.addressof(pyvec)
+        code = "std::pair<std::string, " + vectype + '*>("' + key
+        code += '", reinterpret_cast<' + vectype + f"*>({address_prefix}{vecaddress}))"
+
+        return code
+
+    # Iterate over dictionary, convert numpy arrays to RVecs and put together interpreter code
+    code += ", ".join([write_vec_code(key, value) for key, value in np_dict.items()]) + ")"
+
+    # Create RDataFrame and build Python proxy
+    err = ROOT.gInterpreter.Declare('#include "ROOT/RNumpyDS.hxx"')
+    if not err:
+        raise RuntimeError('Failed to find "ROOT/RNumpyDS.hxx".')
+
+    address = ROOT.gInterpreter.Calc(code)
+    rdf = cppyy.bind_object(address, "ROOT::RDataFrame")
+    ROOT.SetOwnership(rdf, True)
+
+    # Bind pyobject holding adopted memory to the RVec
+    setattr(rdf, "__data__", pyvecs)
+
+    return rdf
```
All of this code does not belong to `_facade.py`, should be moved to some helper module

guitargeek replied:
That's true. I moved it to `_rvec`, `_rtensor`, and `_rdataframe`.

vepadulano commented on bindings/pyroot/pythonizations/python/ROOT/_facade.py:
```diff
@@ -9,13 +9,186 @@
+def _CheckEndianessFromTypestr(typestr):
+    import ROOT
+
+    global byteswap
```
No `global` please

guitargeek replied:
I got rid of it and made it a data member of the facase (for now :))

vepadulano commented on bindings/pyroot/pythonizations/python/ROOT/_facade.py:
```diff
@@ -9,13 +9,186 @@
+        )
+
+
+def _GetCppTypeFromNumpyType(dtype):
+    if dtype == "i4":
+        return "int"
+    elif dtype == "u4":
+        return "unsigned int"
+    elif dtype == "i8":
+        return "Long64_t"
+    elif dtype == "u8":
+        return "ULong64_t"
+    elif dtype == "f4":
+        return "float"
+    elif dtype == "f8":
+        return "double"
+
+    raise RuntimeError("Object not convertible: Python object has unknown data-type '" + dtype + "'.")
```
This is a good chance to convert this function to a Python dictionary which would be much more idiomatic

vepadulano commented on bindings/pyroot/pythonizations/python/ROOT/_facade.py:
```diff
@@ -9,13 +9,186 @@
+    if len(np_dict) == 0:
+        raise RuntimeError("Object not convertible: Dictionary is empty.")
+
+    is_windows = os.name == "nt"
+    address_prefix = "0x" if is_windows else ""
```
```suggestion
    address_prefix = "0x" if platform.system() == "Windows" else ""
```

https://docs.python.org/3/library/platform.html#platform.system

vepadulano reviewed (changes requested):
Thanks! I like the idea, I left some initial comments about things I believe should be improved

guitargeek commented:
Hi @vepadulano, thanks a lot for reviewing this PR! I have addressed your comments, and also migrated the `AsRTensor` function to Python, in the same way as the `AsRVec` function.

The changes after the review comments are for now appended as fixup commits, but at the end of the review I will squash the commits.

hahnjo commented:
@phsft-bot build!

vepadulano commented on bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rvec.py:
```diff
@@ -96,17 +165,15 @@ def get_array_interface(self):
             return {
-                "shape": (size, ),
+                "shape": (size,),
                 "typestr": "{}{}{}".format(endianess, dtype_numpy, dtype_size),
                 "version": 3,
-                "data": (pointer, False)
+                "data": (pointer, False),
```
These cosmetic changes could go a in a separate commit

guitargeek replied:
I have put the cosmetic changes in a separate commit

vepadulano commented on bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rvec.py:
```diff
@@ -96,17 +165,15 @@ def get_array_interface(self):
-    if True in [
-            name.endswith("<{}>".format(dtype)) for dtype in _array_interface_dtype_map
-    ]:
+    if True in [name.endswith("<{}>".format(dtype)) for dtype in _array_interface_dtype_map]:
```
Cosmetic

guitargeek replied:
I have put the cosmetic changes in a separate commit

vepadulano commented on bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:
```diff
@@ -423,3 +423,70 @@ def pythonize_rdataframe(klass):
+    import ROOT
+    import platform
+
+    cppyy = ROOT.module.cppyy
```
Why can't we `import cppyy` directly in this case?

guitargeek replied:
That's true, I think it can be done

vepadulano commented on bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_rtensor.py:
```diff
@@ -9,10 +9,84 @@
+    # Construct an RTensor of the correct data-type
+    prefix = "0x" if platform.system() == "Windows" else ""
+    klassname = f"TMVA::Experimental::RTensor<{cppdtype},std::vector<{cppdtype}>>"
+    code = f"new {klassname}(reinterpret_cast<{cppdtype} *>({prefix}{data}),"
+    code += "{" + ",".join([f"{s}" for s in shape]) + "}, "
+    code += "{" + ",".join([f"{s // dtypesize}" for s in strides]) + "}, "
+    if row_major:
+       code += "TMVA::Experimental::MemoryLayout::RowMajor"
+    else:
+       code += "TMVA::Experimental::MemoryLayout::ColumnMajor"
+    code += ")"
```
Bit of a nitpick, but I believe a multiline f-string would be more idiomatic here

guitargeek replied:
Sure, why not!

vepadulano reviewed (commented):
Thanks a lot! Some minor comments for consideration, I believe this could be already good to merge (I'll check the CI results tomorrow)

guitargeek commented:
Hi @vepadulano, thanks for your follow-up review! I have implemented your comments. Let's hope the CI is green tomorrow morning :+1:

vepadulano commented on bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_rtensor.py:
```diff
@@ -9,10 +9,83 @@
+
+    # Construct an RTensor of the correct data-type
+    prefix = "0x" if platform.system() == "Windows" else ""
+    klassname = f"TMVA::Experimental::RTensor<{cppdtype},std::vector<{cppdtype}>>"
+    address = ROOT.gInterpreter.Calc(
+        f"new {klassname}(reinterpret_cast<{cppdtype} *>({prefix}{data}),"
+        "{" + ",".join([f"{s}" for s in shape]) + "}, "
+        "{" + ",".join([f"{s // dtypesize}" for s in strides]) + "}, "
+        "TMVA::Experimental::MemoryLayout::" + memory_layout + ")"
+    )
+
+    # Bind the object to a Python-side proxy, taking ownership of the underlying C++ object
+    out = cppyy.bind_object(address, klassname)
```
Can we make these lines even more Pythonic? Something like the following (untested pseudocode)
```suggestion
    tensor = TMVA.RTensor[cppdtype, std.vector[cppdtype]](
        cppyy.ll.reinterpret_cast[f"{cppdtype} *"](f"{prefix}{data}"),
        shapetuple,
        memory_layout
    )
```

Then I guess also the ownership will be automatically handled if this works

guitargeek replied:
Thanks a lot, very good suggestions! I implemented them. Let's see what the CI says now.

Unfortunately, I still have a Windows failure that I still don't understand. Maybe it will be magically gone after implementing your suggestions :)

vepadulano commented on bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rvec.py:
```diff
@@ -82,6 +82,75 @@
+        )
+
+    dtype = typestr[1:]
+    cppdtype = _get_cpp_type_from_numpy_type(dtype)
+
+    # Construct an RVec of the correct data-type
+    klassname = "ROOT::VecOps::RVec<" + cppdtype + ">"
+
+    prefix = "0x" if platform.system() == "Windows" else ""
+
+    address = ROOT.gInterpreter.Calc(f"new {klassname}(reinterpret_cast<{cppdtype} *>({prefix}{data}), {size})")
+
+    # Bind the object to a Python-side proxy, taking ownership of the underlying C++ object
+    out = cppyy.bind_object(address, klassname)
+    ROOT.SetOwnership(out, True)
```
Same comment as with `_rtensor`, it would be very nice if we can avoid the extra call to the interpreter

vepadulano commented on bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rvec.py:
```diff
@@ -82,6 +82,61 @@
+    cppdtype = _get_cpp_type_from_numpy_type(dtype)
+
+    # Construct an RVec of the correct data-type
+    out = ROOT.VecOps.RVec[cppdtype](ROOT.module.cppyy.ll.reinterpret_cast[f"{cppdtype} *"](data), [size])
```
Shouldn't this be?
```suggestion
    out = ROOT.VecOps.RVec[cppdtype](ROOT.module.cppyy.ll.reinterpret_cast[f"{cppdtype} *"](data), size)
```

guitargeek replied:
Right, thanks

guitargeek commented:
Sorry I had to update it so many times! ROOT `master` changes so much in PyROOT, I needed to update more than just fixing git conflicts...

vepadulano commented on bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:
```diff
@@ -402,24 +403,89 @@ def pythonize_rdataframe(klass):
+    code += ", ".join([write_vec_code(key, value) for key, value in np_dict.items()]) + ")"
+
+    # Create RDataFrame and build Python proxy
+    err = ROOT.gInterpreter.Declare('#include "ROOT/RNumpyDS.hxx"')
+    if not err:
+        raise RuntimeError('Failed to find "ROOT/RNumpyDS.hxx".')
```
Is this even needed? In a ROOT build with RDataFrame and PyROOT enabled, `RNumpyDS.hxx` should already be there? We could remove another call to the interpreter

vepadulano replied:
Clearly this is much less urgent than other places since it's done only once at construction of the RDataFrame object

guitargeek replied:
Interesting observation, but for now it doesn't work without removing this line. I think this is something for a follow-up PR.

vepadulano commented on bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:
```diff
@@ -402,24 +403,89 @@ def pythonize_rdataframe(klass):
+    rdf = cppyy.bind_object(address, "ROOT::RDataFrame")
+    ROOT.SetOwnership(rdf, True)
+
+    # Bind pyobject holding adopted memory to the RVec
+    setattr(rdf, "__data__", pyvecs)
```
```suggestion
    # Bind pyobject holding adopted memory to the RVec
   rdf.__data__ = pyvecs
```

vepadulano commented on bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rvec.py:
```diff
@@ -82,6 +82,61 @@
+    out = ROOT.VecOps.RVec[cppdtype](ROOT.module.cppyy.ll.reinterpret_cast[f"{cppdtype} *"](data), size)
+
+    # Bind pyobject holding adopted memory to the RVec
+    setattr(out, "__adopted__", arr)
```
```suggestion
    out.__adopted__ = arr
```

vepadulano commented on bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_rtensor.py:
```diff
@@ -7,12 +7,72 @@
+    strides = arr.strides
+
+    # Infer memory layout from strides
+    memory_layout = ROOT.TMVA.Experimental.MemoryLayout.RowMajor
+    if len(strides) > 1 and strides[0] < strides[-1]:
+        memory_layout = ROOT.TMVA.Experimental.MemoryLayout.ColumnMajor
```
Make this a ternary operator assignment maybe

vepadulano commented on bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_rtensor.py:
```diff
@@ -7,12 +7,72 @@
+    )
+
+    # Bind pyobject holding adopted memory to the RTensor
+    setattr(out, "__adopted__", arr)
```
```suggestion
    out.__adopted__ = arr
```

vepadulano reviewed (changes requested):
Super good! A few final comments for consideration before final approval

guitargeek commented:
Hi @vepadulano! I have addressed your latest comments, thanks for the review :+1:

vepadulano reviewed (approved):
Amazing!!! Thanks a lot, let's merge this :rocket:
